torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+) -> PlacementStrategy:
+    """
+    util function to generate a shard strategy on shard_dim
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Shard(shard_dim)]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_replicate_strategy(
+    mesh: DeviceMesh, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a replicate strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Replicate()]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_partial_strategy(mesh: DeviceMesh) -> PlacementStrategy:
+    """
+    util function to generate a partial strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[_Partial()]),
+    )
+
+
+def build_data_parallel_strategies(
+    train_step_graph: GraphModule,
+    num_params: int,
+    num_states: int,
+    mesh: DeviceMesh,
+    batch_dim: int = 0,
+) -> Dict[fx.Node, StrategyList]:
+    """
+    This function loop through the train step graph and build the
+    data parallel strategy for each fx Node
+    """
+    activation_idx = num_params + num_states
+    non_compute_ops = [
+        torch.ops._spmd.tag_grad.default,
+        aten.ones_like.default,
+        aten.t.default,
+        aten.view.default,
+        aten.reshape.default,
+        aten.detach.default,
+        aten.clone.default,
+    ]
+
+    dp_strategy_map = {}
+    batch_dim_analzer = BatchDimAnalyzer(batch_dim)
+    placeholder_idx = 0
+    num_param_grad = 0
+
+    # first we backward propagate to mark the param gradients sharding
+    # with tag_grad node helps and then delete the tag_grad nodes
+    for node in reversed(list(train_step_graph.graph.nodes)):
+        # find a param_grad node via the tagging
+        if node.target == torch.ops._spmd.tag_grad.default:
+            cur_node = node
+            while cur_node.target in non_compute_ops:
+                cur_node = cur_node.args[0]
+                partial_strategy = _gen_partial_strategy(mesh)
+                dp_strategy_map[cur_node] = DataParallelStrategy(
+                    NodeType.GRAD, [partial_strategy]
+                )
+            num_param_grad += 1
+            # remove the tag_grad node from graph
+            node.replace_all_uses_with(node.args[0])
+            train_step_graph.graph.erase_node(node)
+
+            if num_param_grad == num_params:
+                # early break if we have already processed all param_grads
+                break
+
+    train_step_graph.recompile()
+    train_step_graph.print_readable()
+
+    # next we forward propagate to mark all the sharding
+    for node in train_step_graph.graph.nodes:
+        if node.op == "placeholder":
+            if placeholder_idx < num_params:
+                shard_sig = _gen_shard_strategy(mesh, 0)
+                replica_sig = _gen_replicate_strategy(mesh)
```

Comment:
I assume the above two lines mean the param can either be distributed using FSDP (`shard_sig`) and DDP (`replica_sig`)?

For `shard_sig`, do we need to find the best dim to shard in follow up PRs?